### DISCUSSION TITLE
Certificate request with UPN value in SAN

### DISCUSSION THREAD

**Discussion by mb1232142** [strongswan/strongswan#1890] [Q&A]
I want to establish an IPSec tunnel between an router with strongswan and an Fortigate Firewall.
The first step of authentication is an valid certificate, requested by the router and signed by the PKI infrastructure. Due to huge number of routers that will establish IPSec tunnels, the certificate approval needs to be performed automaticly using wildcard.

To ensure security, Fortigate allows the UPN value of the SAN in the certificate to be checked against an LDAP server with the authorized users to perform VPN connections. And UPN is the only value that Fortigate uses to validate against LDAP.

I've checked Strongswan documentation, and didn't find any references on how to fill the certificate request with UPN value. The only reference is to email, that can be similar to UPN, but if is not defined as such is ignored by Fortigate.

Is there any command for that?

Kind regards,

**Comment by tobiasbrunner** [comment]
> The first step of authentication is an valid certificate, requested by the router and signed by the PKI infrastructure.

Could you clarify what you mean. Does that happen outside of IKE?

> Due to huge number of routers that will establish IPSec tunnels, the certificate approval needs to be performed automaticly using wildcard.

What does that mean exactly?

> To ensure security, Fortigate allows the UPN value of the SAN in the certificate to be checked against an LDAP server with the authorized users to perform VPN connections.

What does UPN mean? And what does "UPN value **of the SAN**" mean?

> how to fill the certificate request with UPN value.

What certificate request are you referring to?

**Reply by mb1232142** [reply]
Thank you for your reply.
My post had probably more context than what was needed. So, to rephrase:

In the same certificate that is used to perform IKE authentication, I need to fill the SAN - Subject Alternative Name field with an value named UPN - User Principal Name (see the image below as an example of a certificate with UPN value).
The reason is to implement a validation of allowed users on an LDAP server, as you can see here: https://docs.fortinet.com/document/fortigate/7.4.0/administration-guide/443323/dialup-ipsec-vpn-with-certificate-authentication

The request for that certificate is made by an router with Strongswan, that needs to fill the request with all information needed for IKE authentication, and for my case, also needs to fill the SAN field with that UPN value being something like "Other Name: Principal Name = something@domain .com". Is this possible?

![image](https://github.com/strongswan/strongswan/assets/144354764/6de5cca1-2374-4fec-b933-ff38c83f383b)

**Reply by tobiasbrunner** [reply]
> I need to fill the SAN - Subject Alternative Name field with an value named UPN - User Principal Name

That seems to be some proprietary Microsoft extension (OID is apparently `1.3.6.1.4.1.311.20.2.3`). There is code in the _x509_ plugin that maps such SANs to identities of type RFC822_ADDR (i.e. email addresses). But our identity parser currently can't encode it in DNs or SANs (see [the docs](https://docs.strongswan.org/docs/5.9/config/identityParsing.html)).

> The reason is to implement a validation of allowed users on an LDAP server, as you can see here

According to the [Fortinet docs](https://docs.fortinet.com/document/fortigate/7.4.0/administration-guide/332596), it is possible to specify which SAN is used for this feature (via `set account-key-upn-san`).

> The request for that certificate is made by an router with Strongswan, that needs to fill the request with all information needed for IKE authentication

Do you use strongSwan's [`pki --req`](https://docs.strongswan.org/docs/5.9/pki/pkiReq.html) command to generate a certificate request? Or what do you mean?
